keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,162 @@
+introduces an extra scheduling loop that can be avoided.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades. 
+
+    It is true that pods in Kubernetes don't have sticky IPs by default. In the scenario where the node fails, 
+    we implemented a CNI plugin to ensure that the new pod allocated still get the same ip address.
+
+
+## Goals
+Enable deployment controller to update the containers inside the pods (such as container image, cmd etc.) without tearing down the pods
+
+## Proposal
+
+### API Change
+Augment deployment controller to have a new update strategy type: InPlaceUpdate.
+
+```
+
+type DeploymentStrategy struct {
+	// Type of deployment. Can be "Recreate" or "RollingUpdate" or "InPlaceUpdate". Default is RollingUpdate.
+	// +optional
+	Type DeploymentStrategyType
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update
++ type InPlaceUpdateDeployment struct {
++	 // The number of pods to be updated in each batch.
++	 // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
++	 // Absolute number is calculated from percentage by rounding down.
++	 BatchSize intstr.IntOrString
++ }
+```
+
+User specifies the new update strategy type in the spec and issue the request
+
+### Deployment Controller
+
+#### Update
+Below is the proposed approach for in-place update.
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    // BatchSize is the number of pods to update at each round.
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation etc.
```

Comment:
What happens at this point needs to be spelled out more explicitly.

1. `copyInfo = oldReplicaSet`

What does this actually mean? Are you creating a new replica set in etcd that is essentially a copy of the old that you want to save? If so, what happens when you cannot create the replica set because of quota?

2. `oldReplicaSet = newReplicaSet`

When the old replica set is updated to the new replica set, at that point you have two replica sets that match the deployment spec. You may need to ensure that [`FindNewReplicaSet`](https://github.com/kubernetes/kubernetes/blob/440b2b6ea77e01fb4e2ad99c0b25ac7993e0e00f/pkg/controller/deployment/util/deployment_util.go#L650) handles that case correctly when the new strategy is used.

3. `newReplicaSet = copyInfo`

Assuming you have created the extra copy, are you going to clean it up or continue using it in subsequent rollouts?

Reply:
How about storing the copy as an annotation in `newReplicaSet`?

Reply:
1. ```What does this actually mean? Are you creating a new replica set in etcd that is essentially a copy of the old that you want to save? If so, what happens when you cannot create the replica set because of quota?```
Your proposal to store the copy as an annotation makes sense.
2. ```When the old replica set is updated to the new replica set, at that point you have two replica sets that match the deployment spec. You may need to ensure that FindNewReplicaSet handles that case correctly when the new strategy is used.```
Good point.   Since functions like "FindNewReplicaSet" sorts ReplicaSet by timestamp and assume current ReplicaSet has the latest timestamp, we may also need to swap the timestamp of the oldReplicaSet and new ReplicaSet

3. ```Assuming you have created the extra copy, are you going to clean it up or continue using it in subsequent rollouts?```
Keeping it for the purpose of rollback, the copy basically serves as a historic info for rollback.